## Exploring ebay car sales

This project uses a dataset of cars from a classified section of the German eBay website. I will try and find out information about this data. 

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv",encoding="Latin-1")

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There is a mixture of integer and string data. The columns with integer types are yearOfRegistration, powerPS, monthOfRegistration, nrOfPictures and postalCode. All other columns have string types. There are some null values.

In [3]:
#Changing specific column names as specified using .rename()
new_names = {"yearOfRegistration":"registration_year",
            "monthOfRegistration":"registration_month",
            "notRepairedDamage":"unrepaired_damage",
            "dateCreated":"ad_created"}
autos.rename(new_names,axis=1,inplace=True)

#Changing the rest using a function to convert from camel to snake case
def camel_to_snake(s):
    return ''.join(['_'+c.lower() if c.isupper() else c for c in s]).lstrip('_')

autos.columns = [camel_to_snake(s) for s in autos.columns]

#Changing power_p_s to power_ps
autos.rename({"power_p_s":"power_ps"},axis=1,inplace=True)

#Display column names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [4]:
autos.head()

#Investigating the seller columns
#autos["seller"].value_counts()
#autos.loc[autos["seller"]=='gewerblich']

#Investigating null values
#autos.loc[autos["model"].isnull()]

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

price and odometer are numbers stored as strings (price has a $ symbol, odometer has km at the end). I will clean these columns

nr_of_pictures has 0 for every value and should be removed

seller has 'privat' for every value except one. This value however seems to be legitimate otherwise.

gearbox,model,fuel type,unrepaired damage have null values. I used autos.loc[autos["model"].isnull()] and similar code to investigate these null values. I could not find a pattern from a brief overview.





In [5]:
#Now to clean the price and odometer columns.

autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

autos.head()

autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

/Users/TomMacbook/anaconda3/envs/KV7006/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  This is separate from the ipykernel package so we can avoid doing imports until


In [6]:
autos["odometer_km"].describe()
autos["price"].describe()

autos.sort_values("price")["price"]

18089           0
43923           0
15225           0
29499           0
8445            0
           ...   
27371    12345678
47598    12345678
39377    12345678
42221    27322222
39705    99999999
Name: price, Length: 50000, dtype: int64

The value for odometer_km maxes out at 150000.

There are suspicious prices. There are values 11111111 and 12345678 and 99999999 and 1234566 which seem suspicous and should be removed. It is feasible a car could sell for 10 million or 27 million so I will leave those, although they are very high.

In [7]:
bad_numbers = [11111111,12345678,99999999,1234566]

autos = autos.loc[autos["price"] != 1234566,:]
#I did this with each of the 'bad numbers'. Need to find out how to do it all at once.

autos.sort_values("price")["price"]

1041            0
6459            0
21389           0
21388           0
33243           0
           ...   
39377    12345678
47598    12345678
27371    12345678
42221    27322222
39705    99999999
Name: price, Length: 49999, dtype: int64

In [8]:
autos["odometer_km"].value_counts()
#All seems to be ok here, although the values have clearly been rounded.

150000    32423
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

I had to remove suspicious large values from price. odometer_km seems to be ok but the values have clearly been rounded to the nearest 5000 (if under 10,000), 10,000 (if under 100,000) or the nearest 25,000 (if above 100,000). This is bad data but I'm not sure what I can do about it.

In [9]:
autos["date_crawled"].str[:10].value_counts(normalize=True,
                                           dropna=False).sort_index()
'''
autos["ad_created"].str[:10].value_counts(normalize=True,
                                           dropna=False).sort_index()
autos["last_seen"].str[:10].value_counts(normalize=True,
                                           dropna=False).sort_index()
'''

'\nautos["ad_created"].str[:10].value_counts(normalize=True,\n                                           dropna=False).sort_index()\nautos["last_seen"].str[:10].value_counts(normalize=True,\n                                           dropna=False).sort_index()\n'

In [10]:
autos["registration_year"].describe()

count    49999.000000
mean      2005.073401
std        105.713866
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Outliers in the registration year (years of 1000 and 9999). The date posted seems to rise more in march.

In [11]:
autos["registration_year"].value_counts().sort_index()

1000    1
1001    1
1111    1
1500    1
1800    2
       ..
6200    1
8888    1
9000    2
9996    1
9999    4
Name: registration_year, Length: 97, dtype: int64

It seems values between 1900 to 2016.

In [12]:
autos = autos.loc[autos["registration_year"]<2017,:]
autos = autos.loc[autos["registration_year"]>1900,:]
autos["registration_year"].describe()
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48027 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48027 non-null  object
 1   name                48027 non-null  object
 2   seller              48027 non-null  object
 3   offer_type          48027 non-null  object
 4   price               48027 non-null  int64 
 5   abtest              48027 non-null  object
 6   vehicle_type        44902 non-null  object
 7   registration_year   48027 non-null  int64 
 8   gearbox             45603 non-null  object
 9   power_ps            48027 non-null  int64 
 10  model               45560 non-null  object
 11  odometer_km         48027 non-null  int64 
 12  registration_month  48027 non-null  int64 
 13  fuel_type           44300 non-null  object
 14  brand               48027 non-null  object
 15  unrepaired_damage   39040 non-null  object
 16  ad_created          48

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The incorrect values have been removed.

In [13]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [14]:
autos["brand"].value_counts()
top_20 = autos["brand"].value_counts().sort_values(ascending=False)[:19].index

b_dict = {}
for b in top_20:
    b_list = autos.loc[autos["brand"]==b,:]
    b_mean = b_list["price"].mean()
    b_dict[b] = b_mean
b_dict

{'volkswagen': 6516.457597173145,
 'bmw': 8102.536248343744,
 'opel': 5252.61655437921,
 'mercedes_benz': 30317.447816593885,
 'audi': 9093.65003615329,
 'ford': 7263.015811455847,
 'renault': 2395.4164467897976,
 'peugeot': 3039.4682651622,
 'fiat': 2711.8011272141707,
 'seat': 4296.492554410081,
 'skoda': 6334.91948051948,
 'mazda': 4010.7716643741405,
 'nissan': 4664.891034482758,
 'citroen': 44534.79671150971,
 'smart': 3542.706586826347,
 'toyota': 5115.33388981636,
 'sonstige_autos': 39621.77946768061,
 'hyundai': 5308.53911205074,
 'volvo': 4757.108108108108}

Citroen, mercedes_benz, sonstige_autos are having their mean shifted by extreme values.

I now find the average milage and average price for the top 6 brands.

In [15]:
autos["brand"].value_counts()
top_6 = autos["brand"].value_counts().sort_values(ascending=False)[:5].index

price_dict = {}
for b in top_6:
    b_list = autos.loc[autos["brand"]==b,:]
    b_mean = b_list["price"].mean()
    price_dict[b] = b_mean
    
mileage_dict = {}
for b in top_6:
    b_list = autos.loc[autos["brand"]==b,:]
    b_mean = b_list["odometer_km"].mean()
    mileage_dict[b] = b_mean
    
price_series = pd.Series(price_dict)
mileage_series = pd.Series(mileage_dict)

mileage_series

df = pd.DataFrame(price_series, columns=['mean_price'])
df["mean_mileage"] = mileage_series
df

mean_price   mean_mileage
volkswagen      6516.457597  128730.369062
bmw             8102.536248  132431.383684
opel            5252.616554  129227.141482
mercedes_benz  30317.447817  130860.262009
audi            9093.650036  129287.780188

In [16]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')